tsdb/store.go

```diff
@@ -1709,67 +1709,87 @@ func (s *Store) TagValues(ctx context.Context, auth query.Authorizer, shardIDs [
 			switch e.Op {
 			case influxql.EQ, influxql.NEQ, influxql.EQREGEX, influxql.NEQREGEX:
 				tag, ok := e.LHS.(*influxql.VarRef)
-				if !ok || tag.Val != "_name" {
+				if !ok || tag.Val != measurementKey {
 					return nil
 				}
 			}
 		}
 		return e
 	}), nil)
 
-	filterExpr := influxql.CloneExpr(cond)
-	filterExpr = influxql.Reduce(influxql.RewriteExpr(filterExpr, func(e influxql.Expr) influxql.Expr {
+	is, err := s.indexSetFromShards(ctx, shardIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	names, err := is.MeasurementNamesByExpr(nil, measurementExpr)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.tagValues(ctx, names, is, auth, cond)
+}
+
+// TagValuesByPredicate gets a slice of tag values, provided a flux predicate
+// that has been parsed into an influxql.Expr. It differs from TagValues in that
+// it works with flux predicates.
+func (s *Store) TagValuesByPredicate(ctx context.Context, auth query.Authorizer, shardIDs []uint64, cond influxql.Expr) ([]TagValues, error) {
```

Comment:
This method gets a list of measurement names using the `MeasurementNamesByPredicate` instead of the `MeasurementNamesByExpr` method, and processes the `measurementExpr` a little differently to avoid condensing the AST incorrectly for Flux predicates.

Reply:
Do we 'condense incorrectly' for InfluxQL? Can you expand in the comment on exactly what is different between the flux predicates and InfluxQL predicates, especially since both functions take an influxql.Expr?